### PR TITLE
Add parentheses to import exposing clause if missing

### PR DESCRIPTION
**Pull request by andys8**
Solves #510

### Changes

This PR adds the case that import statements with missing parentheses are allowed and will get parsed successfully, even they're not valid elm syntax. Because this is convenience, I intentionally made the decision not to include line breaks or comments, to keep things simple.

```elm
-- Wrong
import Date exposing Date

-- Right
import Date exposing (Date)
```

I'm trying to learn more about Haskell and am therefore open for hints (also "Allow edits from maintainers"). And a follow up question: Is there code formatting for the Haskell code base? How to use it?

### CODE REVIEW

File: tests/test-files/transform/MissingImportParens.elm [hunk 1]
```diff
@@ -0,0 +1,14 @@
+module Main exposing (x)
+
+import Import.As1 as ImportAs1
+import Import.As2 as ImportAs2 exposing A
+import ImportDotDot exposing ..
+import ImportName exposing A
+import ImportNames exposing A, B, x, y
+import OtherImport1 exposing (A(X, Y), b, z)
+import OtherImport2 exposing (..)
+import OtherImport3
```
Comment (avh4) [inline tests/test-files/transform/MissingImportParens.elm]: The "OtherImport*" lines are testing things that (hopefully!) should be tested in other files (in particular, somewhere in `tests/test-files/good/**`).  I'd probably leave them out here for that reason, though I could be convinced either way on this.

File: tests/test-files/transform/MissingImportParens.elm [hunk 2]
```diff
@@ -0,0 +1,14 @@
+module Main exposing (x)
+
+import Import.As1 as ImportAs1
+import Import.As2 as ImportAs2 exposing A
+import ImportDotDot exposing ..
```
Comment (avh4) [inline tests/test-files/transform/MissingImportParens.elm]: I wonder if we should really allow this one (the `..` without parens) to start with.  (I'm noting it's an explicit line of code that's added to allow this.)  This seems much less likely for people to actually type it this way.  I'd rather avoid adding this on speculation, but if you think this is actually common for people that will be benefiting from this feature, then I guess it's okay.  What do you think?

File: parser/src/Parse/Module.hs [hunk 3]
```diff
@@ -181,7 +181,11 @@ import' =
     exposing =
       do  preExposing <- try (whitespace <* reserved "exposing")
           postExposing <- whitespace
-          imports <- listing detailedListing
+          imports <-
+            choice
+              [ listing detailedListing
+              , listingWithoutParens
```
Comment (avh4) [inline parser/src/Parse/Module.hs]: Should we consider allowing missing parens for all listings, not just ones in the exposing clause of imports?  (I guess that's a question that can always be put off until later.)

File: parser/src/Parse/Module.hs [hunk 4]
```diff
@@ -181,7 +181,11 @@ import' =
     exposing =
       do  preExposing <- try (whitespace <* reserved "exposing")
           postExposing <- whitespace
-          imports <- listing detailedListing
+          imports <-
+            choice
```
Comment (avh4) [inline parser/src/Parse/Module.hs]: Since you asked for feedback:  The thing I'm thinking about when reviewing this code is what potential performance implication this `choice` has (which depends on whether the early choices allow for significant backtracking (with `try`) or not).  In this case, `listing` does `try (char '(')` so I think that will do a maximum of 1 character of backtracking, which should be fine.  (If there were the possibility of significant backtracking, then I'd want to look at ways to rewrite the choice to avoid the backtracking.)

In any case, I think it looks good for this particular change.

File: tests/test-files/transform/MissingImportParens.elm [hunk 5]
```diff
@@ -0,0 +1,14 @@
...
+import ImportName exposing A
+import ImportNames exposing A, B, x, y
```
Comment (avh4) [inline tests/test-files/transform/MissingImportParens.elm]: I think there are multiline cases to consider.

What will happen with 

```
import ModuleName exposing A,

f = ()
```
(I guess in this specific example, parsing would fail... but specifically, I'm wondering are there any cases where a typo might cause the new parsing branches to end up successfully consuming a bunch of code it shouldn't, and lead to an unexpected formatting result?)

Do we want to allow things like
```
import ModuleName exposing
   A, B
```

Either we should be conservative, and only this feature when everything's on a single line, or make sure we have test cases for the multiline variants we can think of, and think through whether there are any cases that could lead to a bad result.

Comment (avh4) [reply tests/test-files/transform/MissingImportParens.elm]: Oh, you noted in the original post that you intentionally decided not to include line breaks... but `detailedListing` uses `commaSepBy1'`, which I think will consume comments, whitespace, and newlines immediately after a comma (maybe I'm wrong, though).  Did you try it to see what happens?

**Comment by avh4** [comment]
Yeah, I haven't been using a code formatter for the haskell code.  I try to format it like Elm code (by hand) for new code, but I'm not really worried about it.

File: tests/test-files/transform/MissingImportParens.elm (hunk 5 above)
Comment (andys8) [reply tests/test-files/transform/MissingImportParens.elm]: `detailedListing` uses ` commaSep1'`.
https://github.com/avh4/elm-format/blob/master/parser/src/Parse/Helpers.hs#L246-L248
and `spaceySepBy1'`
https://github.com/avh4/elm-format/blob/master/parser/src/Parse/Helpers.hs#L236-L238
which uses `spaceySepBy1''`
https://github.com/avh4/elm-format/blob/master/parser/src/Parse/Helpers.hs#L214-L232

What's your idea to solve this?

Comment (andys8) [reply tests/test-files/transform/MissingImportParens.elm]: The example above with the line breaks should and is not working. I'm adding a test file for bad tests. But to be honest, I can't foresee the edge cases and their consequences.

File: parser/src/Parse/Module.hs (hunk 4 above)
Comment (andys8) [reply parser/src/Parse/Module.hs]: Great, thanks for the info. I also thought about the order, and having the standard case at first, failing at the missing "(".

File: parser/src/Parse/Module.hs (hunk 3 above)
Comment (andys8) [reply parser/src/Parse/Module.hs]: So, `listing` is also used for type declarations and tuples? I'd be careful, in this case.

File: tests/test-files/transform/MissingImportParens.elm (hunk 2 above)
Comment (andys8) [reply tests/test-files/transform/MissingImportParens.elm]: There are arguments for both sides. There is an extra branch to support it and we're not sure, how likely it is for people to use. On the other side: Having a named value working, but ".." not, is inconsistent. Why is it like this and how should the user know?

Since it would work right now, I'd go and support it.

File: tests/test-files/transform/MissingImportParens.elm (hunk 1 above)
Comment (andys8) [reply tests/test-files/transform/MissingImportParens.elm]: You're right. My intention to have the good cases, was not to break them. While developing I had cases with a wrong parser, where the good case failed (because too much was parsed?). I'll reduce the standard cases, but I think it makes sense to have them here, to test the case, that there is a mix of different possibilities, and they they're working together in the same file. Also adding the merge case.

**Comment by andys8** [comment]
Thanks a lot for the input! I rebased and added changes.

We could think about formatting the haskell code base and integrate tooling in the future, especially since this is a formatting tool's code base :)

File: tests/test-files/bad/MissingImportParens.elm [hunk 6]
```diff
@@ -0,0 +1,4 @@
+module Main exposing (..)
+
+import ImportModuleWithoutParensWithLineBreak exposing
+   A, B
```
Comment (avh4) [inline tests/test-files/bad/MissingImportParens.elm]: Elm files are invalid if they contain no definitions...  Does this test case still fail if you add `x = ()` (on its own line) at the end of the file?

Comment (andys8) [reply tests/test-files/bad/MissingImportParens.elm]: You're right. This was the case and not the actual parsing to fail. It looks to my like parsing line breaks is working. Using the existing implementation I'm not sure what to expect regarding line breaks. Do you mind having a look into it? Is the current state of this PR working as we expect it?

**Comment by andys8** [comment]
Sounds promising. Thanks @avh4

File: tests/test-files/transform/MissingImportParens.elm (hunk 5 above)
Comment (avh4) [reply tests/test-files/transform/MissingImportParens.elm]: Okay, playing with this a bit more, if there is a trailing comma, then it will consume the next token, but any subsequent elm definition must start at the beginning of a line, so if an extra token was consumed, then it won't be the beginning of the line anymore and will fail to parse.  So I think this seems safe. 👍
